File: Java Projects/Numeric Matrix Processor (Medium)/Stage 4/Processor.java
package processor;

import java.util.Scanner;
import java.util.Arrays;

public class Processor {
    private Scanner scanner;

    public Processor() {
        scanner = new Scanner(System.in);
        Matrix.setScanner(scanner);
    }

    public void execute() {
        executionLoop:
        while (true) {
            System.out.println("1. Add matrices");
            System.out.println("2. Multiply matrix by a constant");
            System.out.println("3. Multiply matrices");
            System.out.println("4. Transpose matrix");
            System.out.println("0. Exit");
            System.out.print("Your choice: ");
            String command = scanner.nextLine();
            try {
                switch (command) {
                    case "1":
                        addMatrices();
                        break;
                    case "2":
                        scalarMultiplication();
                        break;
                    case "3":
                        matrixMultiplication();
                        break;
                    case "4":
                        matrixTranspose();
                        break;
                    case "0":
                        break executionLoop;
                }
            } catch (IllegalArgumentException e) {
                System.out.println("The operation cannot be performed.");
            }
            System.out.println();
        }
    }

    private void addMatrices() throws IllegalArgumentException {
        System.out.print("Enter size of first matrix: ");
        int[] dims = getDimensions();
        System.out.println("Enter first matrix: ");
        Matrix a = Matrix.getMatrix(dims[0], dims[1]);
        System.out.print("Enter size of second matrix: ");
        dims = getDimensions();
        System.out.println("Enter second matrix: ");
        Matrix b = Matrix.getMatrix(dims[0], dims[1]);
        System.out.println("The result is: ");
        Matrix.add(a, b).printMatrix();
    }

    private void scalarMultiplication() {
        System.out.print("Enter size of matrix: ");
        int[] dims = getDimensions();
        System.out.println("Enter matrix: ");
        Matrix a = Matrix.getMatrix(dims[0], dims[1]);
        System.out.print("Enter constant: ");
        double constant = Double.parseDouble(scanner.next());
        System.out.println("The result is: ");
        Matrix.scalarMultiply(constant, a).printMatrix();
    }

    private void matrixMultiplication() throws IllegalArgumentException {
        System.out.print("Enter size of first matrix: ");
        int[] dims = getDimensions();
        System.out.println("Enter first matrix: ");
        Matrix a = Matrix.getMatrix(dims[0], dims[1]);
        System.out.print("Enter size of second matrix: ");
        dims = getDimensions();
        System.out.println("Enter second matrix: ");
        Matrix b = Matrix.getMatrix(dims[0], dims[1]);
        System.out.println("The result is: ");
        Matrix.multiply(a, b).printMatrix();
    }

    private void matrixTranspose() {
        System.out.println();
        System.out.println("1. Main diagonal");
        System.out.println("2. Side diagonal");
        System.out.println("3. Vertical line");
        System.out.println("4. Horizontal line");
        System.out.print("your choice: ");
        String command = scanner.nextLine();
        System.out.print("Enter matrix size: ");
        int[] dims = getDimensions();
        System.out.println("Enter matrix: ");
        Matrix a = Matrix.getMatrix(dims[0], dims[1]);
        System.out.println("The result is: ");
        switch (command) {
            case "1":
                Matrix.transposeMainDiagonal(a).printMatrix();
                break;
            case "2":
                Matrix.transposeSideDiagonal(a).printMatrix();
                break;
            case "3":
                Matrix.transposeVertical(a).printMatrix();
                break;
            case "4":
                Matrix.transposeHorizontal(a).printMatrix();
                break;
        }
    }

    private int[] getDimensions() {
        return Arrays.stream(scanner.nextLine().split(" ")).mapToInt(Integer::parseInt).toArray();
    }
}

File: Python Projects/Numeric Matrix Processor (Medium)/Stage 5/processor.py
from Matrix import Matrix


def load_matrix(n_rows):
    fields = [[int(float(i)) if float(i).is_integer() else float(i) for i in input().split()] for _ in range(n_rows)]
    return Matrix(fields)


def add_matrices():
    a, b = load_two_matrices()
    return a + b


def scalar_multiply():
    matrix = load_one_matrix()
    scalar = float(input("Enter constant: "))
    scalar = int(scalar) if scalar.is_integer() else scalar
    return matrix * scalar


def matrix_multiply():
    a, b = load_two_matrices()
    return a @ b


def load_one_matrix():
    row = int(input("Enter size of matrix:").split()[0])
    print("Enter matrix:")
    return load_matrix(row)



def load_two_matrices():
    rows = int(input("Enter size of first matrix:").split()[0])
    print("Enter first matrix:")
    a = load_matrix(rows)
    rows = int(input("Enter size of second matrix:").split()[0])
    print("Enter second matrix:")
    b = load_matrix(rows)
    return a, b


def transpose_matrix():
    print("1. Main diagonal")
    print("2. Side diagonal")
    print("3. Vertical line")
    print("4. Horizontal line")
    choice = int(input("Your choice: "))
    matrix = load_one_matrix()
    if choice == 1:
        result = matrix.transpose_main_diagonal()
    elif choice == 2:
        result = matrix.transpose_side_diagonal()
    elif choice == 3:
        result = matrix.transpose_vertical_line()
    elif choice == 4:
        result = matrix.transpose_horizontal_line()
    print("The result is: ")
    print(result)


def determinant():
    matrix = load_one_matrix()
    print("The result is: ")
    print(matrix.determinant())


def menu():
    while True:
        print("1. Add matrices")
        print("2. Multiply matrix by a constant")
        print("3. Multiply matrices")
        print("4. Transpose matrix")
        print("5. Calculate a determinant")
        print("0. Exit")
        choice = int(input("Your choice: "))
        if choice == 0:
            break
        elif choice == 1:
            try:
                result = add_matrices()
                print("The result is: ")
                print(result)
            except Exception:
                print("The operation cannot be performed")
        elif choice == 2:
            result = scalar_multiply()
            print("The result is: ")
            print(result)
        elif choice == 3:
            try:
                result = matrix_multiply()
                print("The result is: ")
                print(result)
            except Exception:
                print("The operation cannot be performed")
        elif choice == 4:
            print()
            transpose_matrix()
        elif choice == 5:
            determinant()
        print()


menu()